renter/renter_test.go

```diff
@@ -0,0 +1,729 @@
+	contract, _, err := session.FormContract(renterPriv, types.Siacoins(5000), types.Siacoins(10000), renterNode.c.Tip().Height+200, settings, renterNode.w, renterNode.tp)
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	// upload data to the contract
+	vc := renterNode.c.TipContext()
+	contractPayment := session.PayByContract(&contract, renterPriv, renterPriv.PublicKey())
+	hostSettings, err = session.RegisterSettings(context.Background(), contractPayment)
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	buf := bytes.NewBuffer(nil)
+	builder := rhp.NewProgramBuilder(hostSettings, buf, contract.Revision.WindowEnd-vc.Index.Height)
+	sectors := make([][rhp.SectorSize]byte, 10)
+	for i := 0; i < len(sectors); i++ {
+		frand.Read(sectors[i][:256])
+		builder.AddAppendSectorInstruction(&sectors[i], true)
+	}
+	cost := builder.Cost()
+	instructions, requiresContract, requiresFinalization, err := builder.Program()
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	budget := cost.BaseCost.Add(cost.StorageCost).Add(settings.DownloadBandwidthPrice.Mul64(1 << 12)).Add(settings.UploadBandwidthPrice.Mul64(rhp.SectorSize * 11))
+	if _, err := session.FundAccount(context.Background(), renterPriv.PublicKey(), budget, contractPayment); err != nil {
+		b.Fatal(err)
+	}
+
+	payment := session.PayByEphemeralAccount(renterPriv.PublicKey(), renterPriv, vc.Index.Height+10)
+	err = session.ExecuteProgram(renter.Program{
+		Instructions: instructions,
+		Budget:       budget,
+
+		RequiresContract:     requiresContract,
+		RequiresFinalization: requiresFinalization,
+		Contract:             &contract,
+		RenterKey:            renterPriv,
+	}, buf.Bytes(), payment, renter.NoopOnOutput)
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	buf.Reset()
+	builder = rhp.NewProgramBuilder(hostSettings, buf, contract.Revision.WindowEnd-vc.Index.Height)
+	roots := make([]types.Hash256, 256)
+	for i := 0; i < len(roots); i++ {
+		roots[i] = rhp.SectorRoot(&sectors[frand.Intn(len(sectors))])
+		if err := builder.AddReadSectorInstruction(roots[i], 0, rhp.SectorSize, false); err != nil {
+			b.Fatal(err)
+		}
+	}
+
+	cost = builder.Cost()
+	instructions, requiresContract, requiresFinalization, err = builder.Program()
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	budget = cost.BaseCost.Add(settings.DownloadBandwidthPrice.Mul64(rhp.SectorSize * 300))
+	if _, err := session.FundAccount(context.Background(), renterPriv.PublicKey(), budget, contractPayment); err != nil {
+		b.Fatal(err)
+	}
+
+	// execute download program
+	b.ResetTimer()
+	b.ReportAllocs()
+	b.SetBytes(256 * rhp.SectorSize)
+	err = session.ExecuteProgram(renter.Program{
+		Instructions: instructions,
+		Budget:       budget,
+
+		RequiresContract:     requiresContract,
+		RequiresFinalization: requiresFinalization,
+		Contract:             &contract,
+		RenterKey:            renterPriv,
+	}, buf.Bytes(), payment, renter.NoopOnOutput)
+	if err != nil {
+		b.Fatal(err)
+	}
+}
+
+func BenchmarkUploadProgram(b *testing.B) {
```

Comment:
doesn't look like this benchmark uses `b.N` anywhere, so it won't run properly. You can either upload in a loop (`b.N` times), or upload `b.N` bytes.